Pandas Accessing Data


Review Data

Let's load in the data using `read_csv` and quickly review it using `head()`.


In [1]:
# Importing Libraries
import pandas as pd
from datasets import load_dataset
import matplotlib.pyplot as plt

# Loading Data
dataset = load_dataset('lukebarousse/data_jobs')
df = dataset['train'].to_pandas()

# Data Cleanup
df.job_posted_date = pd.to_datetime(df.job_posted_date)

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 785741 entries, 0 to 785740
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   job_title_short        785741 non-null  object        
 1   job_title              785740 non-null  object        
 2   job_location           784696 non-null  object        
 3   job_via                785733 non-null  object        
 4   job_schedule_type      773074 non-null  object        
 5   job_work_from_home     785741 non-null  bool          
 6   search_location        785741 non-null  object        
 7   job_posted_date        785741 non-null  datetime64[ns]
 8   job_no_degree_mention  785741 non-null  bool          
 9   job_health_insurance   785741 non-null  bool          
 10  job_country            785692 non-null  object        
 11  salary_rate            33067 non-null   object        
 12  salary_year_avg        22003 non-null   floa

In [3]:
df.head()

job_title_short                                          job_title  \
0  Senior Data Engineer  Senior Clinical Data Engineer / Principal Clin...   
1          Data Analyst                                       Data Analyst   
2         Data Engineer  Data Engineer/Scientist/Analyst, Mid or Senior...   
3         Data Engineer  LEAD ENGINEER - PRINCIPAL ANALYST - PRINCIPAL ...   
4         Data Engineer                             Data Engineer- Sr Jobs   

                   job_location             job_via job_schedule_type  \
0                 Watertown, CT     via Work Nearby         Full-time   
1  Guadalajara, Jalisco, Mexico    via BeBee México         Full-time   
2               Berlin, Germany        via LinkedIn         Full-time   
3               San Antonio, TX   via Diversity.com         Full-time   
4                Washington, DC  via Clearance Jobs         Full-time   

   job_work_from_home       search_location     job_posted_date  \
0               False  Texas, United States 2023-06-16 13:44:15   
1               False                Mexico 2023-01-14 13:18:07   
2               False               Germany 2023-10-10 13:14:55   
3               False  Texas, United States 2023-07-04 13:01:41   
4               False                 Sudan 2023-08-07 14:29:36   

   job_no_degree_mention  job_health_insurance    job_country salary_rate  \
0                  False                 False  United States        None   
1                  False                 False         Mexico        None   
2                  False                 False        Germany        None   
3                   True                 False  United States        None   
4                  False                 False          Sudan        None   

   salary_year_avg  salary_hour_avg                  company_name  \
0              NaN              NaN          Boehringer Ingelheim   
1              NaN              NaN    Hewlett Packard Enterprise   
2              NaN              NaN      ALPHA Augmented Services   
3              NaN              NaN  Southwest Research Institute   
4              NaN              NaN               Kristina Daniel   

                                          job_skills  \
0                                               None   
1  ['r', 'python', 'sql', 'nosql', 'power bi', 't...   
2  ['python', 'sql', 'c#', 'azure', 'airflow', 'd...   
3  ['python', 'c++', 'java', 'matlab', 'aws', 'te...   
4  ['bash', 'python', 'oracle', 'aws', 'ansible',...   

                                     job_type_skills  
0                                               None  
1  {'analyst_tools': ['power bi', 'tableau'], 'pr...  
2  {'analyst_tools': ['dax'], 'cloud': ['azure'],...  
3  {'cloud': ['aws'], 'libraries': ['tensorflow',...  
4  {'cloud': ['oracle', 'aws'], 'other': ['ansibl...

We learned how to get rows using `iloc[]` before. But we can do a lot more with it. We can actually get rows _and_ columns.


When to Choose `loc` vs. `iloc`?

Use `loc`:

When your DataFrame has meaningful index labels (e.g., 'a', 'b', or dates).
When filtering data with conditions (e.g., df['Age'] > 30).
When selecting specific columns by name.

Use `iloc`:

When you need to select data by position (e.g., first 5 rows, columns 2 and 3).
When the index labels are irrelevant or you’re working with numerical indices.
When programmatically accessing data by position (e.g., in a loop).

Notes

Performance: Both are efficient, but `iloc` can be slightly faster for large DataFrames when using integer positions.
Errors: Using `loc` with a non-existent label raises a KeyError. Using `iloc` with an out-of-bounds position raises an IndexError.
Mixed Indexing: Avoid mixing `loc` and `iloc` in the same operation to prevent confusion.


iloc

Notes

- `df.iloc[]`: Select rows and columns by position.

Examples
Using `iloc` let's:

1. Get the first row (index 0).
2. Get the `job_skills` element of the 3rd (index at 2) row.
3. Get the `job_skills` (index 15) and `job_type_skills` (index 16) for the third (index 2) and fourth (index 3) rows.
4. Get the first 12 rows of the DataFrame.
5. Get the first five columns of the DataFrame and all of the rows.

For this we'll need to know the index numbers for our DataFrame.


1. Get the first row.


In [4]:
df.iloc[0]

job_title_short                                       Senior Data Engineer
job_title                Senior Clinical Data Engineer / Principal Clin...
job_location                                                 Watertown, CT
job_via                                                    via Work Nearby
job_schedule_type                                                Full-time
job_work_from_home                                                   False
search_location                                       Texas, United States
job_posted_date                                        2023-06-16 13:44:15
job_no_degree_mention                                                False
job_health_insurance                                                 False
job_country                                                  United States
salary_rate                                                           None
salary_year_avg                                                        NaN
salary_hour_avg          

2. Get the `job_skills` element of the 3rd (index at 2) row.


In [5]:
df.iloc[0][15]

/var/folders/4c/d3ftmjdx6ldd8ylkqq_nd7n40000gn/T/ipykernel_49328/2485859661.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  df.iloc[0][15]


Note: Use `df.iloc[0, 15]` instead of `df.iloc[0][15]` to ensure future compatibility with pandas.

The use of chained indexing like `df.iloc[0][15]` is being deprecated in pandas, as it may lead to ambiguous behavior between positional and label-based access in future versions. By using `df.iloc[0, 15]`, you directly specify the position of the data you want to access, which is clearer and avoids potential future errors when pandas changes how integer keys are interpreted in series indexing.

So we should instead write:


In [6]:
print(df.iloc[0,15])

None


3. Get the `job_skills` (index 15) and `job_type_skills` (index 16) for the third (index 2) and fourth (index 3) rows.
   - To get third (index 2) and fourth (index 3) rows: `[2,3]`
   - To get `job_skills` and `job_type_skills` which are index 15 and index 16 retrospectively: `[15,16]`
   - Then you put those two into a list itself to get everything between these two: `df.iloc[[2,3],[15,16]]`


In [7]:
df.iloc[[2, 3], [15, 16]]

job_skills  \
2  ['python', 'sql', 'c#', 'azure', 'airflow', 'd...   
3  ['python', 'c++', 'java', 'matlab', 'aws', 'te...   

                                     job_type_skills  
2  {'analyst_tools': ['dax'], 'cloud': ['azure'],...  
3  {'cloud': ['aws'], 'libraries': ['tensorflow',...

Preview

Below are a few more examples of what `iloc` can do. Pay close attention to these because we'll use them later.


4. Get the first 10 rows of the DataFrame.


In [8]:
df.iloc[:9]

job_title_short                                          job_title  \
0  Senior Data Engineer  Senior Clinical Data Engineer / Principal Clin...   
1          Data Analyst                                       Data Analyst   
2         Data Engineer  Data Engineer/Scientist/Analyst, Mid or Senior...   
3         Data Engineer  LEAD ENGINEER - PRINCIPAL ANALYST - PRINCIPAL ...   
4         Data Engineer                             Data Engineer- Sr Jobs   
5         Data Engineer                                  GCP Data Engineer   
6  Senior Data Engineer                  Senior Data Engineer  - GCP Cloud   
7         Data Engineer                                      Data Engineer   
8      Business Analyst           Technology & Operations Business Analyst   

                   job_location             job_via         job_schedule_type  \
0                 Watertown, CT     via Work Nearby                 Full-time   
1  Guadalajara, Jalisco, Mexico    via BeBee México                 Full-time   
2               Berlin, Germany        via LinkedIn                 Full-time   
3               San Antonio, TX   via Diversity.com                 Full-time   
4                Washington, DC  via Clearance Jobs                 Full-time   
5                      Anywhere    via ZipRecruiter  Contractor and Temp work   
6                  Dearborn, MI        via LinkedIn                 Full-time   
7                      Anywhere        via LinkedIn                 Full-time   
8           Copenhagen, Denmark     via Trabajo.org                 Full-time   

   job_work_from_home         search_location     job_posted_date  \
0               False    Texas, United States 2023-06-16 13:44:15   
1               False                  Mexico 2023-01-14 13:18:07   
2               False                 Germany 2023-10-10 13:14:55   
3               False    Texas, United States 2023-07-04 13:01:41   
4               False                   Sudan 2023-08-07 14:29:36   
5                True                 Georgia 2023-11-07 14:01:59   
6               False  Florida, United States 2023-03-27 13:18:18   
7                True                 Romania 2023-12-07 13:40:49   
8               False                 Denmark 2023-06-05 13:44:34   

   job_no_degree_mention  job_health_insurance    job_country salary_rate  \
0                  False                 False  United States        None   
1                  False                 False         Mexico        None   
2                  False                 False        Germany        None   
3                   True                 False  United States        None   
4                  False                 False          Sudan        None   
5                  False                 False  United States        None   
6                  False                 False  United States        None   
7                  False                 False        Romania        None   
8                  False                 False        Denmark        None   

   salary_year_avg  salary_hour_avg                   company_name  \
0              NaN              NaN           Boehringer Ingelheim   
1              NaN              NaN     Hewlett Packard Enterprise   
2              NaN              NaN       ALPHA Augmented Services   
3              NaN              NaN   Southwest Research Institute   
4              NaN              NaN                Kristina Daniel   
5              NaN              NaN                smart folks inc   
6              NaN              NaN  Miracle Software Systems, Inc   
7              NaN              NaN                          Zitec   
8              NaN              NaN                         Hempel   

                                          job_skills  \
0                                               None   
1  ['r', 'python', 'sql', 'nosql', 'power bi', 't...   
2  ['python', 'sql', 'c#', 'azure', 'airflow', 'd...   
3  ['python', 'c++', 'java', 

4. Get the first 10 rows of the DataFrame.


In [9]:
df.iloc[:, : 5]

job_title_short  \
0       Senior Data Engineer   
1               Data Analyst   
2              Data Engineer   
3              Data Engineer   
4              Data Engineer   
...                      ...   
785736     Software Engineer   
785737          Data Analyst   
785738      Business Analyst   
785739         Data Engineer   
785740     Software Engineer   

                                                job_title  \
0       Senior Clinical Data Engineer / Principal Clin...   
1                                            Data Analyst   
2       Data Engineer/Scientist/Analyst, Mid or Senior...   
3       LEAD ENGINEER - PRINCIPAL ANALYST - PRINCIPAL ...   
4                                  Data Engineer- Sr Jobs   
...                                                   ...   
785736                                    DevOps Engineer   
785737                                   CRM Data Analyst   
785738                     Commercial Analyst - Start Now   
785739  Principal Associate, Data Engineer (Remote-Eli...   
785740                                 AWS System Analyst   

                               job_location                    job_via  \
0                             Watertown, CT            via Work Nearby   
1              Guadalajara, Jalisco, Mexico           via BeBee México   
2                           Berlin, Germany               via LinkedIn   
3                           San Antonio, TX          via Diversity.com   
4                            Washington, DC         via Clearance Jobs   
...                                     ...                        ...   
785736                            Singapura        melalui Trabajo.org   
785737                   Bad Rodach, Jerman  melalui BeBee Deutschland   
785738                             Malaysia           melalui Ricebowl   
785739  Newark, New Jersey, Amerika Serikat        melalui Recruit.net   
785740                                India             melalui Trigyn   

       job_schedule_type  
0              Full-time  
1              Full-time  
2              Full-time  
3              Full-time  
4              Full-time  
...                  ...  
785736   Pekerjaan tetap  
785737   Pekerjaan tetap  
785738   Pekerjaan tetap  
785739   Pekerjaan tetap  
785740   Pekerjaan tetap  

[785741 rows x 5 columns]

loc

Notes

- `df.loc[]`: Select rows and columns by position or label.
- Similar to `df.iloc[]` except we can use labels instead.

Example

Let's get the same columns we did before but with `loc` instead. Which uses column and row labels. These are:

1. Get the first row (index 0).
2. Get the first 10 rows of `job_skills` and `job_type_skills`.
3. Get the first 5 columns and rows 10-20.
4. Get the first 12 rows of the DataFrame.
5. Get the first 5 columns of the DataFrame and all the rows.


1. Get the first row. This remains the same because the row doesn't have a label.


In [10]:
df.loc[0]

job_title_short                                       Senior Data Engineer
job_title                Senior Clinical Data Engineer / Principal Clin...
job_location                                                 Watertown, CT
job_via                                                    via Work Nearby
job_schedule_type                                                Full-time
job_work_from_home                                                   False
search_location                                       Texas, United States
job_posted_date                                        2023-06-16 13:44:15
job_no_degree_mention                                                False
job_health_insurance                                                 False
job_country                                                  United States
salary_rate                                                           None
salary_year_avg                                                        NaN
salary_hour_avg          

2. Get the first 10 rows of `job_skills` and `job_type_skills`.


In [11]:
df.loc[: 9, ['job_skills', 'job_type_skills']]

job_skills  \
0                                               None   
1  ['r', 'python', 'sql', 'nosql', 'power bi', 't...   
2  ['python', 'sql', 'c#', 'azure', 'airflow', 'd...   
3  ['python', 'c++', 'java', 'matlab', 'aws', 'te...   
4  ['bash', 'python', 'oracle', 'aws', 'ansible',...   
5                           ['python', 'sql', 'gcp']   
6  ['sql', 'python', 'java', 'sql server', 'gcp',...   
7  ['sql', 'nosql', 'gcp', 'azure', 'aws', 'bigqu...   
8                ['excel', 'powerpoint', 'power bi']   
9  ['sql', 'python', 'r', 'mongodb', 'mongodb', '...   

                                     job_type_skills  
0                                               None  
1  {'analyst_tools': ['power bi', 'tableau'], 'pr...  
2  {'analyst_tools': ['dax'], 'cloud': ['azure'],...  
3  {'cloud': ['aws'], 'libraries': ['tensorflow',...  
4  {'cloud': ['oracle', 'aws'], 'other': ['ansibl...  
5  {'cloud': ['gcp'], 'programming': ['python', '...  
6  {'cloud': ['gcp', 'bigquery'], 'databases': ['...  
7  {'cloud': ['gcp', 'azure', 'aws', 'bigquery', ...  
8  {'analyst_tools': ['excel', 'powerpoint', 'pow...  
9  {'analyst_tools': ['excel'], 'cloud': ['azure'...

In [12]:
df.loc[9][["job_skills", "job_type_skills"]]

job_skills         ['sql', 'python', 'r', 'mongodb', 'mongodb', '...
job_type_skills    {'analyst_tools': ['excel'], 'cloud': ['azure'...
Name: 9, dtype: object

3. Get the first 5 columns and rows 10-20.


In [13]:
df.loc[10:20,'job_title_short':'job_work_from_home']

job_title_short                                    job_title  \
10         Data Engineer         Erfahrene*r Data Engineer*in (m/w/d)   
11         Data Engineer                                Data Engineer   
12          Data Analyst  Stagiaire Data Analyst (H/F) - Lyon (69006)   
13  Senior Data Engineer                         Senior Data Engineer   
14         Data Engineer                                Data Engineer   
15         Data Engineer                                Data Engineer   
16         Data Engineer                            Big Data Engineer   
17        Data Scientist                       Data Science Team Lead   
18          Data Analyst                                 Data Analyst   
19        Data Scientist                               Data Scientist   
20        Data Scientist                               Data Scientist   

                                       job_location  \
10                               Basel, Switzerland   
11               Colorado Springs, CO   (+3 others)   
12                                     Lyon, France   
13                                     New York, NY   
14  Rio de Janeiro, State of Rio de Janeiro, Brazil   
15                                     Aberdeen, UK   
16                         Pune, Maharashtra, India   
17                                       Laurel, MD   
18                                   Warsaw, Poland   
19                         Pune, Maharashtra, India   
20                                Zaventem, Belgium   

                                              job_via  \
10                                       via LinkedIn   
11                                       via The Muse   
12                                       via Jobijoba   
13                                       via LinkedIn   
14                                          via BeBee   
15                                       via LinkedIn   
16                                       via LinkedIn   
17  via APL Careers - Johns Hopkins University App...   
18                              via Praca Trabajo.org   
19                                         via Indeed   
20                                 via BeBee Belgique   

           job_schedule_type  job_work_from_home  
10                 Full-time               False  
11                 Full-time               False  
12  Full-time and Internship               False  
13                 Full-time               False  
14                 Full-time               False  
15                 Temp work               False  
16                 Full-time               False  
17                 Full-time               False  
18                 Full-time               False  
19                 Full-time               False  
20                 Full-time               False

Pandas Data Cleaning


Handling Missing Data


Review

This is what we learned in the basics section, this is just a refresher of how we've handled null values before. Feel free to skip this.

Notes

- `df.dropna()`: Drop missing values.

Examples

Here we are only drop values if all of their values are missing.


In [14]:
df_cleaned = df.dropna(how='all')
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 785741 entries, 0 to 785740
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   job_title_short        785741 non-null  object        
 1   job_title              785740 non-null  object        
 2   job_location           784696 non-null  object        
 3   job_via                785733 non-null  object        
 4   job_schedule_type      773074 non-null  object        
 5   job_work_from_home     785741 non-null  bool          
 6   search_location        785741 non-null  object        
 7   job_posted_date        785741 non-null  datetime64[ns]
 8   job_no_degree_mention  785741 non-null  bool          
 9   job_health_insurance   785741 non-null  bool          
 10  job_country            785692 non-null  object        
 11  salary_rate            33067 non-null   object        
 12  salary_year_avg        22003 non-null   floa

Right now all we can do is drop values if they're missing. But that's not useful right now because our DataFrame didn't have any.

So, what if we wanted to fill the missing values with something else? This is expecially useful so we don't run into errors when dealing with NaN values.


Fillna

Notes

- `df.fillna()`: Fill missing values

Examples

Let's fill in instances where there's no salary info (aka these columns have NaN values `salary_rate`, `salary_year_avg`, `salary_hour_avg`) with 0.


We're going to look at a few rows in these 3 columns right now, so we can compare what we've done before to after.

We'll use `iloc` to look at the first 10 rows `:10` and the salary information rows `11:14`.


In [15]:
df.iloc[:10, 11:14]

salary_rate  salary_year_avg  salary_hour_avg
0        None              NaN              NaN
1        None              NaN              NaN
2        None              NaN              NaN
3        None              NaN              NaN
4        None              NaN              NaN
5        None              NaN              NaN
6        None              NaN              NaN
7        None              NaN              NaN
8        None              NaN              NaN
9        None              NaN              NaN

We fill the values for the 3 columns with 0 using `fillna()`.


In [16]:
fill_values = ['salary_rate', 'salary_year_avg', 'salary_hour_avg']
df_filled = df_cleaned.fillna(0)

df_filled.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 785741 entries, 0 to 785740
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   job_title_short        785741 non-null  object        
 1   job_title              785741 non-null  object        
 2   job_location           785741 non-null  object        
 3   job_via                785741 non-null  object        
 4   job_schedule_type      785741 non-null  object        
 5   job_work_from_home     785741 non-null  bool          
 6   search_location        785741 non-null  object        
 7   job_posted_date        785741 non-null  datetime64[ns]
 8   job_no_degree_mention  785741 non-null  bool          
 9   job_health_insurance   785741 non-null  bool          
 10  job_country            785741 non-null  object        
 11  salary_rate            785741 non-null  object        
 12  salary_year_avg        785741 non-null  floa

Now if we compare the results using `iloc` again on our new DataFrame. We see that the previous NaN values in the columns ( `salary_rate`, `salary_year_avg`, `salary_hour_avg`) have been replaced with 0.


In [17]:
df_filled.iloc[:10, 11: 14]

salary_rate  salary_year_avg  salary_hour_avg
0           0              0.0              0.0
1           0              0.0              0.0
2           0              0.0              0.0
3           0              0.0              0.0
4           0              0.0              0.0
5           0              0.0              0.0
6           0              0.0              0.0
7           0              0.0              0.0
8           0              0.0              0.0
9           0              0.0              0.0

Drop Duplicates

Notes

- `drop_duplicates()`: Remove duplicate rows.
- Analysts will often need to clean up data and one of the most common issues we run into is duplicate values.

Examples

Now that we've dealt with NaN values. Let's continue cleaning the data by removing any duplicate rows.


In [18]:
df_unique = df_filled.drop_duplicates()

df_unique.info()

<class 'pandas.core.frame.DataFrame'>
Index: 785640 entries, 0 to 785740
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   job_title_short        785640 non-null  object        
 1   job_title              785640 non-null  object        
 2   job_location           785640 non-null  object        
 3   job_via                785640 non-null  object        
 4   job_schedule_type      785640 non-null  object        
 5   job_work_from_home     785640 non-null  bool          
 6   search_location        785640 non-null  object        
 7   job_posted_date        785640 non-null  datetime64[ns]
 8   job_no_degree_mention  785640 non-null  bool          
 9   job_health_insurance   785640 non-null  bool          
 10  job_country            785640 non-null  object        
 11  salary_rate            785640 non-null  object        
 12  salary_year_avg        785640 non-null  float64  

If you compare that with the original DataFrame which had 785741 entries. This new DataFrame `df_unique` has 785640 entries. It removed 101 entries.

Now let's see what would happen if we tried to remove duplicates from `job_title`.


In [19]:
df_unique = df_filled.drop_duplicates(subset=['job_title'])

df_unique.info()

<class 'pandas.core.frame.DataFrame'>
Index: 234674 entries, 0 to 785740
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   job_title_short        234674 non-null  object        
 1   job_title              234674 non-null  object        
 2   job_location           234674 non-null  object        
 3   job_via                234674 non-null  object        
 4   job_schedule_type      234674 non-null  object        
 5   job_work_from_home     234674 non-null  bool          
 6   search_location        234674 non-null  object        
 7   job_posted_date        234674 non-null  datetime64[ns]
 8   job_no_degree_mention  234674 non-null  bool          
 9   job_health_insurance   234674 non-null  bool          
 10  job_country            234674 non-null  object        
 11  salary_rate            234674 non-null  object        
 12  salary_year_avg        234674 non-null  float64  

If we look at it now. It looks like we removed quite a few rows. Now all of these rows have unique `job_title`s.


In [20]:
df_unique.head(10)

job_title_short                                          job_title  \
0  Senior Data Engineer  Senior Clinical Data Engineer / Principal Clin...   
1          Data Analyst                                       Data Analyst   
2         Data Engineer  Data Engineer/Scientist/Analyst, Mid or Senior...   
3         Data Engineer  LEAD ENGINEER - PRINCIPAL ANALYST - PRINCIPAL ...   
4         Data Engineer                             Data Engineer- Sr Jobs   
5         Data Engineer                                  GCP Data Engineer   
6  Senior Data Engineer                  Senior Data Engineer  - GCP Cloud   
7         Data Engineer                                      Data Engineer   
8      Business Analyst           Technology & Operations Business Analyst   
9        Data Scientist                                  Data Scientist II   

                   job_location             job_via         job_schedule_type  \
0                 Watertown, CT     via Work Nearby                 Full-time   
1  Guadalajara, Jalisco, Mexico    via BeBee México                 Full-time   
2               Berlin, Germany        via LinkedIn                 Full-time   
3               San Antonio, TX   via Diversity.com                 Full-time   
4                Washington, DC  via Clearance Jobs                 Full-time   
5                      Anywhere    via ZipRecruiter  Contractor and Temp work   
6                  Dearborn, MI        via LinkedIn                 Full-time   
7                      Anywhere        via LinkedIn                 Full-time   
8           Copenhagen, Denmark     via Trabajo.org                 Full-time   
9                      Anywhere    via ZipRecruiter                 Full-time   

   job_work_from_home          search_location     job_posted_date  \
0               False     Texas, United States 2023-06-16 13:44:15   
1               False                   Mexico 2023-01-14 13:18:07   
2               False                  Germany 2023-10-10 13:14:55   
3               False     Texas, United States 2023-07-04 13:01:41   
4               False                    Sudan 2023-08-07 14:29:36   
5                True                  Georgia 2023-11-07 14:01:59   
6               False   Florida, United States 2023-03-27 13:18:18   
7                True                  Romania 2023-12-07 13:40:49   
8               False                  Denmark 2023-06-05 13:44:34   
9                True  New York, United States 2023-04-23 13:02:57   

   job_no_degree_mention  job_health_insurance    job_country salary_rate  \
0                  False                 False  United States           0   
1                  False                 False         Mexico           0   
2                  False                 False        Germany           0   
3                   True                 False  United States           0   
4                  False                 False          Sudan           0   
5                  False                 False  United States           0   
6                  False                 False  United States           0   
7                  False                 False        Romania           0   
8                  False                 False        Denmark           0   
9                  False                 False  United States           0   

   salary_year_avg  salary_hour_avg                   company_name  \
0              0.0              0.0           Boehringer Ingelheim   
1              0.0              0.0     Hewlett Packard Enterprise   
2              0.0              0.0       ALPHA Augmented Services   
3              0.0              0.0   Southwest Research Institute   
4              0.0              0.0                Kristina Daniel   
5              0.0              0.0                smart folks inc   
6              0.0              0.0  Miracle Software Systems, Inc   
7              0.0              0.0                          Zitec   
8       

For our example we don't really need to remove any duplicates right now, but it's important to understand the concept.


Pandas Data Management


In [21]:
# DataFrame Copy
df_original = df.copy()

Copy


Recall from the last lesson, when we filled in missing values for median salary.

Here let's make a new dataframe `df_altered` and only make changes to it.


In [22]:
# Create new dataframe
df_altered = df_original

df_altered.loc[:5, 'salary_year_avg']

0   NaN
1   NaN
2   NaN
3   NaN
4   NaN
5   NaN
Name: salary_year_avg, dtype: float64

In [23]:
# Calculating the median salary
median_salary = df_altered.salary_year_avg.median()

# Filling the missing values with the median salary
df_altered.salary_year_avg = df_altered.loc[:, 'salary_year_avg'].fillna(median_salary)

Now let's inspect the altered DataFrame.


In [24]:
df_altered.loc[:5, 'salary_year_avg']

0    115000.0
1    115000.0
2    115000.0
3    115000.0
4    115000.0
5    115000.0
Name: salary_year_avg, dtype: float64

That was good...

But what about the original...


In [25]:
df_original.loc[:5,'salary_year_avg']

0    115000.0
1    115000.0
2    115000.0
3    115000.0
4    115000.0
5    115000.0
Name: salary_year_avg, dtype: float64

Holdup!! How the heck did `df_original` get altered!?!

Well both the variables of `df_original` and `df_altered` are referencing the same DataFrame.


In [26]:
print('ID of df_original:               ', id(df_original))
print('ID of df_altered:                ', id(df_altered))
print('Are the two dataframes the same? ', id(df_original) == id(df_altered))

ID of df_original:                6114796704
ID of df_altered:                 6114796704
Are the two dataframes the same?  True


Instead we can use the .copy() method

- `copy()`: Copy a DataFrame.


In [27]:
df_original = df.copy()
df_altered = df_original.copy()

print('ID of df_original:               ', id(df_original))
print('ID of df_altered:                ', id(df_altered))
print('Are the two dataframes the same? ', id(df_original) == id(df_altered))

ID of df_original:                6114798720
ID of df_altered:                 6114798272
Are the two dataframes the same?  False


Now when we do this same operation:


In [28]:
# Calculating the median salary
median_salary = df_altered['salary_year_avg'].median()

# Filling the missing values with the median salary
df_altered['salary_year_avg'] = df_altered['salary_year_avg'].fillna(median_salary)

df_altered.loc[:5,'salary_year_avg']

0    115000.0
1    115000.0
2    115000.0
3    115000.0
4    115000.0
5    115000.0
Name: salary_year_avg, dtype: float64

The original dataframe doesn't get altered!


In [29]:
df_original.loc[:5,'salary_year_avg']

0   NaN
1   NaN
2   NaN
3   NaN
4   NaN
5   NaN
Name: salary_year_avg, dtype: float64

Now that we've created a copy of our data, we want to start our analysis. But if we have a large set of data we only want to take a subset of data to make it more manageable. We can use `sample()` to get a random sample of the data.


Sample

Notes

- `sample()`: Random sample of items.

Examples

Let's get a random sample of the data. You could get a sample with a fixed row number.


In [30]:
df.sample(n = 5)

job_title_short                             job_title  \
271509   Data Engineer                 Data Engineer (m/w/d)   
672176    Data Analyst                          Data Analyst   
126911  Data Scientist                Data Scientist - Brand   
132923    Data Analyst  Data Analyst - Full-time / Part-time   
730532  Data Scientist                    Data Scientist h/f   

              job_location                  job_via job_schedule_type  \
271509  Stuttgart, Germany                 via XING         Full-time   
672176     Vienna, Austria                via BeBee         Full-time   
126911              Canada            via AngelList         Full-time   
132923    Jacksonville, FL             via Snagajob         Full-time   
730532      Nantes, France  via Emplois Trabajo.org         Full-time   

        job_work_from_home         search_location     job_posted_date  \
271509               False                 Germany 2023-02-20 07:17:14   
672176               False                 Austria 2023-05-31 11:01:50   
126911               False                  Canada 2023-03-01 08:27:55   
132923               False  Florida, United States 2023-02-08 08:02:13   
730532               False                  France 2023-01-19 21:56:43   

        job_no_degree_mention  job_health_insurance    job_country  \
271509                   True                 False        Germany   
672176                  False                 False        Austria   
126911                  False                 False         Canada   
132923                  False                 False  United States   
730532                  False                 False         France   

       salary_rate  salary_year_avg  salary_hour_avg         company_name  \
271509        None              NaN              NaN              Akkodis   
672176        None              NaN              NaN  Barmherzige Brueder   
126911        None              NaN              NaN                Faire   
132923        None              NaN              NaN          Robert Half   
730532        None              NaN              NaN               AVISIA   

                                         job_skills  \
271509       ['python', 'c', 'sql', 'aws', 'azure']   
672176                             ['sql', 'excel']   
126911                            ['python', 'sql']   
132923                              ['spreadsheet']   
730532  ['sas', 'sas', 'python', 'hadoop', 'spark']   

                                          job_type_skills  
271509  {'cloud': ['aws', 'azure'], 'programming': ['p...  
672176  {'analyst_tools': ['excel'], 'programming': ['...  
126911                 {'programming': ['python', 'sql']}  
132923                 {'analyst_tools': ['spreadsheet']}  
730532  {'analyst_tools': ['sas'], 'libraries': ['hado...

Or you can randomly select a fraction of the data (e.g., 10% of the rows), with or without replacement.


In [31]:
df.sample(frac = 0.1, replace = False)

job_title_short                                          job_title  \
664784    Data Scientist  Analytics Education and Implementation Specialist   
188803     Data Engineer                                  Big Data Engineer   
89599       Data Analyst                            Data Analyst Consultant   
128887     Data Engineer                        Data Visualization Engineer   
340863  Business Analyst                                            Analyst   
...                  ...                                                ...   
579273     Data Engineer                           Data Engineer (Contract)   
210368     Data Engineer                          Data Engineer (Snowflake)   
648333  Business Analyst                      Verification Business Analyst   
125746     Data Engineer                                      Data Engineer   
212807      Data Analyst               Data Analytics Training & Internship   

                  job_location                  job_via  \
664784               Tampa, FL         via ZipRecruiter   
188803        Toulouse, France  via Emplois Trabajo.org   
89599          Porto, Portugal              via EMPREGO   
128887     Panama City, Panama         via BeBee Panamá   
340863               Australia        via Jobrapido.com   
...                        ...                      ...   
579273        Yerevan, Armenia     via LinkedIn Armenia   
210368                Anywhere             via LinkedIn   
648333  The Hague, Netherlands                via BeBee   
125746        Florham Park, NJ   via CAIA Career Center   
212807                Anywhere             via LinkedIn   

               job_schedule_type  job_work_from_home         search_location  \
664784                 Full-time               False  Florida, United States   
188803                 Full-time               False                  France   
89599                  Full-time               False                Portugal   
128887                 Full-time               False                  Panama   
340863                 Full-time               False               Australia   
...                          ...                 ...                     ...   
579273  Full-time and Contractor               False                 Armenia   
210368                Contractor                True                 Georgia   
648333                 Temp work               False             Netherlands   
125746                 Full-time               False                 Georgia   
212807                      None                True               Australia   

           job_posted_date  job_no_degree_mention  job_health_insurance  \
664784 2023-12-13 10:02:59                  False                 False   
188803 2023-03-30 15:34:46                  False                 False   
89599  2023-03-24 00:29:04                  False                 False   
128887 2023-09-28 08:48:23                   True                 False   
340863 2023-01-02 23:31:25                  False                 False   
...                    ...                    ...                   ...   
579273 2023-12-28 10:52:03                  False                 False   
210368 2023-09-19 22:35:43                  False                 False   
648333 2023-10-06 12:17:12                  False                 False   
125746 2023-05-10 08:38:07                  False                 False   
212807 2023-03-20 22:20:24                  False                 False   

          job_country salary_rate  salary_year_avg  salary_hour_avg  \
664784  United States        None              NaN              NaN   
188803         France        None              NaN              NaN   
89599        Portugal        None              NaN              NaN   
128887         Panama        None              NaN              NaN   
340863      Australia        None              NaN              NaN   
...               ...         ...              ...             

Now you can analyze these subsets of data.
